pkg/cloud/services/s3/s3.go

```diff
@@ -0,0 +1,284 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package s3
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"path"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/aws/aws-sdk-go/service/sts/stsiface"
+	"github.com/pkg/errors"
+
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/scope"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/hash"
+)
+
+// Service holds a collection of interfaces.
+// The interfaces are broken down like this to group functions together.
+// One alternative is to have a large list of functions from the ec2 client.
+type Service struct {
+	scope     scope.S3Scope
+	S3Client  s3iface.S3API
+	STSClient stsiface.STSAPI
+}
+
+// NewService returns a new service given the api clients.
+func NewService(s3Scope scope.S3Scope) *Service {
+	s3Client := scope.NewS3Client(s3Scope, s3Scope, s3Scope, s3Scope.InfraCluster())
+	STSClient := scope.NewSTSClient(s3Scope, s3Scope, s3Scope, s3Scope.InfraCluster())
+
+	return &Service{
+		scope:     s3Scope,
+		S3Client:  s3Client,
+		STSClient: STSClient,
+	}
+}
+
+func (s *Service) ReconcileBucket() error {
+	if !s.bucketManagementEnabled() {
+		return nil
+	}
+
+	bucketName, err := s.bucketName()
+	if err != nil {
+		return errors.Wrap(err, "generating bucket name")
+	}
+
+	if err := s.createBucketIfNotExist(bucketName); err != nil {
+		return errors.Wrap(err, "ensuring bucket exists")
+	}
+
+	if err := s.ensureBucketPolicy(bucketName); err != nil {
+		return errors.Wrap(err, "ensuring bucket policy")
+	}
+
+	return nil
+}
+
+func (s *Service) DeleteBucket() error {
+	if !s.bucketManagementEnabled() {
+		return nil
+	}
+
+	bucketName, err := s.bucketName()
+	if err != nil {
+		return errors.Wrap(err, "generating bucket name")
+	}
+
+	s.scope.Info("Deleting S3 Bucket", "name", bucketName)
+
+	if _, err := s.S3Client.DeleteBucket(&s3.DeleteBucketInput{
+		Bucket: aws.String(bucketName),
+	}); err != nil {
+		return errors.Wrap(err, "deleting S3 bucket")
+	}
+
+	return nil
+}
+
+func (s *Service) Create(m *scope.MachineScope, data []byte) (string, error) {
+	if !s.bucketManagementEnabled() {
+		return "", errors.New("requested object creation but bucket management is not enabled")
+	}
+
+	if m == nil {
+		return "", errors.New("machine scope can't be nil")
+	}
+
+	if len(data) == 0 {
+		return "", errors.New("got empty data")
+	}
+
+	bucket, err := s.bucketName()
+	if err != nil {
+		return "", errors.Wrap(err, "generating bucket name")
+	}
+	key := s.bootstrapDataKey(m)
+
+	if _, err := s.S3Client.PutObject(&s3.PutObjectInput{
+		Body:   aws.ReadSeekCloser(bytes.NewReader(data)),
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key),
+	}); err != nil {
+		return "", errors.Wrap(err, "putting object")
+	}
```

Comment:
Do we want this as part of this PR or maybe it could be implemented in another one? This one is already large and I'd try to avoid making it even bigger if possible.

Reply:
We already did, so marking as resolved.